llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp

```diff
@@ -4181,11 +4181,26 @@ void SelectionDAGBuilder::visitShuffleVector(const User &I) {
     return;
   }
 
-  // We can't use either concat vectors or extract subvectors so fall back to
+  // If the target prefers, emit a padded shuffle vector at the
+  // source operand width, then extract the original destination
+  // type.
+  if (!TLI.shouldScalarizeLengthDecreasingShuffle()) {
```

Comment:
SelectionDAGBuilder shouldn't be making transformations and should reliably translate what the IR was

Reply:
The problem in this case is that there is no direct translation.  IR supports mixed length shuffles; DAG does not.  I happen to think the padding approach is "closer" to the just translate your asking for, but that's definite an opinion not objective fact.

Reply:
> . IR supports mixed length shuffles; DAG does not.

Then it should. There is a complexity cost to structural deviations in the second IR